packages/vite/src/node/packages.ts

```diff
@@ -163,3 +163,40 @@ export function watchPackageDataPlugin(config: ResolvedConfig): Plugin {
     }
   }
 }
+
+export function findPackageJson(dir: string): string | null {
+  // Stop looking at node_modules directory.
+  if (path.basename(dir) === 'node_modules') {
+    return null
+  }
+  const pkgPath = path.join(dir, 'package.json')
+  if (fs.existsSync(pkgPath)) {
+    return pkgPath
+  }
+  const parentDir = path.dirname(dir)
+  return parentDir !== dir ? findPackageJson(parentDir) : null
+}
+
+const workspaceRootFiles = ['lerna.json', 'pnpm-workspace.yaml', '.git']
+
+export function isWorkspaceRoot(
```

Comment:
Instead of using `searchWorkspaceRoot`, I added this function, which takes advantage of the `packageCache`. It also only checks the given directory, instead of an upward traversal.